File: py/Largest Rectangle in Histogram.py
class Solution:
    # @param height, a list of integer
    # @return an integer
    
    def largestRectangleArea(self, height):
        stack = []
        
        if height == []:
            return 0
        
        height.insert(0,0)
        height.append(0)
        
        iret = 0
        top = -1
        
        for i in range(len(height)):
            if top == -1:
                stack.append(i)
                top += 1
                continue
            
            if height[stack[top]] <= height[i]:
                stack.append(i)
                top += 1
            else:
                while top != -1 and height[stack[top]] > height[i]:
                    icur = stack.pop()
                    top -= 1
                    if (i - stack[top] - 1) * height[icur] > iret:
                        iret = (i - stack[top] - 1) * height[icur]
                stack.append(i)
                top += 1
                        
        return iret

File: cpp/digit_problem/Largest Rectangle in Histogram.cpp
class Solution {
public:
    int largestRectangleArea(vector<int> &height) {
        stack<int> num;
        vector<int>::iterator i = height.begin();
        
        height.insert(i,1,0);
        height.push_back(0);
        
        num.push(0);
        
        int iret = 0;
        
        for(int i = 1;i < height.size();i++)
        {
            if(height[i] < height[num.top()])
            {
                while(num.size() != 0 && height[i] < height[num.top()])
                {
                    int cur = num.top();
                    num.pop();
                    if(iret < (i - num.top() - 1) * height[cur])
                    {
                        iret = (i - num.top() - 1) * height[cur];
                    }
                }
                
            }
            num.push(i);
        }
        
        return iret;
    }
};